File: libs/java/messaging/src/main/java/arda/messaging/consumer/IdempotencyHandler.java
package arda.messaging.consumer;

import reactor.core.publisher.Mono;

/**
 * Interface for handling idempotency to prevent duplicate processing of messages.
 */
public interface IdempotencyHandler {
    /**
     * Check if the message has been processed.
     * @param key Unique key for the message (e.g. Event ID)
     * @return Mono<Boolean> true if already processed, false otherwise
     */
    Mono<Boolean> isProcessed(String key);

    /**
     * Mark the message as processed.
     */
    Mono<Void> markAsProcessed(String key);
}

File: arda-be-java/libs/messaging/src/main/kotlin/arda/messaging/consumer/IdempotencyHandler.kt
package arda.messaging.consumer

import reactor.core.publisher.Mono

/**
 * Interface for handling idempotency to prevent duplicate processing of messages.
 */
interface IdempotencyHandler {
    /**
     * Check if the message has been processed.
     * @param key Unique key for the message (e.g. Event ID)
     * @return Mono<Boolean> true if already processed, false otherwise
     */
    fun isProcessed(key: String): Mono<Boolean>

    /**
     * Mark the message as processed.
     */
    fun markAsProcessed(key: String): Mono<Void>
}
